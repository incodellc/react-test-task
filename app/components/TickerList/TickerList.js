import React from 'react';
import PropTypes from 'prop-types';

import {Ticker} from '../Ticker/Ticker';


const TickerList = ({tickers}) => {
    return (
        <div className="table-responsive-lg">
                <table className="table">
                    <thead>
                    <tr>
                        <th scope="col">#</th>
                        <th scope="col">Ticker</th>
                        <th scope="col">Exchange</th>
                        <th scope="col">Price</th>
                        <th scope="col">Change</th>
                        <th scope="col">Change percent</th>
                        <th scope="col">Last trade time</th>
                        <th scope="col">Dividend</th>
                        <th scope="col">Yield</th>
                    </tr>
                    </thead>
                    <tbody>
                    {tickers.map((item, idx) => {
                        const isPositiveDynamic = !tickers[idx - 1] ? null : item.price >= tickers[idx - 1].price;
                        const className = isPositiveDynamic ? 'arrow-up' : 'arrow-down';
                        return(
                            <Ticker key={idx} id={idx + 1} item={item} className={className}/>
                        );
                    }).reverse()}
                    </tbody>
                </table>
            </div>
    );
};

TickerList.propTypes = {
    tickers: PropTypes.array.isRequired,
};

export default TickerList;

