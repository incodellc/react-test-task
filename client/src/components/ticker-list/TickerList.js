import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { toggleTicker } from '../../store/actions/action-creators';
import { tickersSelector } from '../../store/selectors';
import { List } from 'semantic-ui-react';
import { TickerListItem } from '../../components';

const TickerList = () => {
    const dispatch = useDispatch();
    const tickers = useSelector(tickersSelector);
    const toggleTickerHandler = useCallback((e, { id }) => dispatch(toggleTicker(id)), [dispatch]);

    return (
        <List 
            data-testid='ticker-list'
            celled 
            selection 
            verticalAlign='middle'
        >
            {
                tickers.map(ticker => (
                    <TickerListItem
                        key={ticker.tickerId}
                        ticker={ticker}
                        onTickerClick={toggleTickerHandler}
                    />
                ))
            }
        </List>
    );
};

export default TickerList;