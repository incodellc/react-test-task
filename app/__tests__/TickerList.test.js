import React from 'react'
import { shallow } from 'enzyme'
import TickerList from '../components/TickerList/TickerList';

describe('TickerList test', () => {
    const props = {
        tickers: [
            {
                'ticker': 'AAPL',
                'exchange': 'NASDAQ',
                'price': '116.60',
                'change': '-0.46',
                'change_percent': '-0.39',
                'last_trade_time': 'Oct 21, 4:00PM EDT',
                'dividend': '0.57',
                'yield': '1.96'
            },
            {
                'ticker': 'AAPL',
                'exchange': 'NASDAQ',
                'price': '16.60',
                'change': '-0.36',
                'change_percent': '-0.29',
                'last_trade_time': 'Oct 21, 3:00PM EDT',
                'dividend': '0.55',
                'yield': '1.01'
            }
        ],
    };
    const tickerList = shallow(<TickerList {...props} />);

    it('should renders properly', () => {
        expect(tickerList).toMatchSnapshot();
    });

    it('should renders 9 table-heads (<th>)', () => {
        expect(tickerList.find('th')).toHaveLength(9);
    });

    it('should renders 2 tickers', () => {
        expect(tickerList.find('Ticker')).toHaveLength(2);
    });
});
