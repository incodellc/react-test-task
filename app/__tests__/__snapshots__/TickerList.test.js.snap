// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TickerList test should renders properly 1`] = `
<div
  className="table-responsive-lg"
>
  <table
    className="table"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          #
        </th>
        <th
          scope="col"
        >
          Ticker
        </th>
        <th
          scope="col"
        >
          Exchange
        </th>
        <th
          scope="col"
        >
          Price
        </th>
        <th
          scope="col"
        >
          Change
        </th>
        <th
          scope="col"
        >
          Change percent
        </th>
        <th
          scope="col"
        >
          Last trade time
        </th>
        <th
          scope="col"
        >
          Dividend
        </th>
        <th
          scope="col"
        >
          Yield
        </th>
      </tr>
    </thead>
    <tbody>
      <Ticker
        className="arrow-up"
        id={2}
        item={
          Object {
            "change": "-0.36",
            "change_percent": "-0.29",
            "dividend": "0.55",
            "exchange": "NASDAQ",
            "last_trade_time": "Oct 21, 3:00PM EDT",
            "price": "16.60",
            "ticker": "AAPL",
            "yield": "1.01",
          }
        }
        key="1"
      />
      <Ticker
        className="arrow-down"
        id={1}
        item={
          Object {
            "change": "-0.46",
            "change_percent": "-0.39",
            "dividend": "0.57",
            "exchange": "NASDAQ",
            "last_trade_time": "Oct 21, 4:00PM EDT",
            "price": "116.60",
            "ticker": "AAPL",
            "yield": "1.96",
          }
        }
        key="0"
      />
    </tbody>
  </table>
</div>
`;
